### PR TITLE
Multi tracking file support

### LINKED ISSUE #287: Queen's Feature Request - Toggling Between Tracking Files
When comparing tracking data, such as evaluating someone else's tracking or analyzing bead tracking, the ability to toggle between different tracking files would be very advantageous. This process currently involves manually loading each file individually and then having load a new trial, along with the appropriate filter settings. By enabling a toggling mechanism, we can quickly switch between different tracking datasets, facilitating more efficient comparisons and analysis. This streamlined approach reduces the hassle of repeatedly loading trials and loading filtering settings, making it easier to identify discrepancies and validate results.

### PR DESCRIPTION
**Pull request by NicerNewerCar**
Adds support for loading multiple tracking files
Can use the shortcuts `J` and `K` to swap between loaded tracking files.
A "tracking set" is a context for all loaded volumes, so by adding a single extra tracking set you are able to load two tracking files for each volume.
Closes #287 
switching_tras

### CODE REVIEW

**Comment by DajungYoon** [comment]
@NicerNewerCar Would it be possible to display anywhere the name of the tracking file that you are looking at?

File: autoscoper/src/ui/AutoscoperMainWindow.cpp [hunk 1]
```diff
@@ -2078,6 +2078,30 @@ void AutoscoperMainWindow::on_toolButtonTrackCurrent_clicked()
    tracking_dialog->retrack();
    }*/
 
+void AutoscoperMainWindow::on_toolButtonPreviousCurveSet_clicked()
+{
+  if (tracker)
+    tracker->trial()->decrementActiveCurve();
```
Comment (jcfr) [inline autoscoper/src/ui/AutoscoperMainWindow.cpp:2084]:
```suggestion
  if (tracker) {
    tracker->trial()->decrementActiveCurve();
  }
```

File: libautoscoper/src/VolumeTransform.hpp [hunk 2]
```diff
@@ -57,16 +56,28 @@ class VolumeTransform
...
-  KeyCurve<float> x_curve;
-  KeyCurve<float> y_curve;
-  KeyCurve<float> z_curve;
-  KeyCurve<Quatf> quat_curve;
+  KeyCurve<float>* getXCurve() { return x_curves[activeCurve]; }
+  KeyCurve<float>* getYCurve() { return y_curves[activeCurve]; }
+  KeyCurve<float>* getZCurve() { return z_curves[activeCurve]; }
+  KeyCurve<Quatf>* getQuatCurve() { return quat_curves[activeCurve]; }
+
+  void addBlankCurveSet();
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.hpp:67]: Add `numberOfCurveSets() const` function returning size of `x_curves`

File: libautoscoper/src/VolumeTransform.hpp [hunk 3]
```diff
@@ -57,16 +56,28 @@ class VolumeTransform
...
+  void incrementActiveCurve();
+  void decrementActiveCurve();
 
   // CoordFrame volumeTrans; //FromWorldToVolume
   CoordFrame volumeMatrix; // FromWorldToPivot
+
+private:
+  unsigned int activeCurve;
+  unsigned int numCurves = 0;
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.hpp:76]:
```suggestion
```

Remove and prefer using `numberOfCurveSets`

File: libautoscoper/src/VolumeTransform.cpp [hunk 4]
```diff
@@ -2,13 +2,47 @@
...
+  KeyCurve<float>* y_curve = new KeyCurve<float>(KeyCurve<float>::Y_CURVE);
+  KeyCurve<float>* z_curve = new KeyCurve<float>(KeyCurve<float>::Z_CURVE);
+  KeyCurve<Quatf>* quat_curve = new KeyCurve<Quatf>(KeyCurve<Quatf>::QUAT_CURVE);
+
+  x_curves.push_back(x_curve);
+  y_curves.push_back(y_curve);
+  z_curves.push_back(z_curve);
+  quat_curves.push_back(quat_curve);
+
+  numCurves++;
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:18]:
```suggestion
```

File: libautoscoper/src/VolumeTransform.cpp [hunk 5]
```diff
@@ -2,13 +2,47 @@
 #include "VolumeTransform.hpp"
 
 namespace xromm {
-VolumeTransform::VolumeTransform()
+
+void VolumeTransform::addBlankCurveSet()
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:6]: -> `addCurveSet`

File: libautoscoper/src/VolumeTransform.cpp [hunk 6]
```diff
@@ -2,13 +2,47 @@
...
+  x_curves.push_back(x_curve);
+  y_curves.push_back(y_curve);
+  z_curves.push_back(z_curve);
+  quat_curves.push_back(quat_curve);
+
+  numCurves++;
+  activeCurve = numCurves - 1;
+}
+
+void VolumeTransform::incrementActiveCurve()
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:22]: `incrementActiveCurve` -> `setCurrentCurveSetToNext`

File: libautoscoper/src/VolumeTransform.cpp [hunk 7]
```diff
@@ -2,13 +2,47 @@
...
+  activeCurve++;
+  if (activeCurve >= numCurves) {
+    activeCurve = 0;
+  }
 }
 
-VolumeTransform::~VolumeTransform() {}
-} // namespace xromm
\ No newline at end of file
+void VolumeTransform::decrementActiveCurve()
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:31]: `decrementActiveCurve` -> `setCurrentCurveSetToPrevious`

File: libautoscoper/src/VolumeTransform.cpp [hunk 8]
```diff
@@ -2,13 +2,47 @@
...
+{
+  // Decrement the activeCurve unless it's at 0, then wrap it to the highest index
+  if (activeCurve == 0) {
+    activeCurve = numCurves - 1;
+  } else {
+    activeCurve--;
+  }
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:38]:
```suggestion
  // Decrement the current curveSet or wrap it to the highest index
  currentCurveSet--;
  if (currentCurveSet < 0) {
    currentCurveSet = numberOfCurveSets() - 1;
  }
```

File: libautoscoper/src/VolumeTransform.cpp [hunk 9]
```diff
@@ -2,13 +2,47 @@
...
 {
-  x_curve.type = KeyCurve<float>::X_CURVE;
-  y_curve.type = KeyCurve<float>::Y_CURVE;
-  z_curve.type = KeyCurve<float>::Z_CURVE;
-  quat_curve.type = KeyCurve<Quatf>::QUAT_CURVE;
+  // Increment the activeCurve, unless it goes above the numCurves, then wrap it back to 0
+  activeCurve++;
+  if (activeCurve >= numCurves) {
+    activeCurve = 0;
+  }
```
Comment (jcfr) [inline libautoscoper/src/VolumeTransform.cpp:28]:
```suggestion
  // Increment the current curveSet or wrap it back to 0
  currentCurveSet++;
  if (currentCurveSet >= numberOfCurveSets()) {
    currentCurveSet = 0;
  }
```

**Review by jcfr — CHANGES_REQUESTED** [review]
Overall the changes are sensible. Before integrating, I suggest to perform few API name changes.

File: autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui [hunk 10]
```diff
@@ -384,6 +384,49 @@ Dialog</string>
...
+          <enum>Qt::Vertical</enum>
+         </property>
+        </widget>
+       </item>
+       <item>
+        <widget class="QToolButton" name="toolButtonPreviousCurveSet">
+         <property name="text">
+          <string>Previous
+Tracking
+Set (J)</string>
```
Comment (jcfr) [inline autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui:399]: Is the line wrapping required ?

Comment (NicerNewerCar) [reply autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui:399]: Unfortunately, it is:
```xml
          <string>Previous 
            Tracking
            Set (J)
          </string>
```
image

Comment (jcfr) [reply autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui:399]: In that case, it should be possible to use html, see https://doc.qt.io/qt-5/richtext-html-subset.html

```
<center><p>Previous</p><p>Tracking</p><p>Set (J)</p></center>
```

Comment (jcfr) [reply autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui:399]: HTML is not supported in this case, instead using a line break should ensure the text is centered:
```
Previous\nTracking\nSet (J)
```

Comment (jcfr) [reply autoscoper/src/ui/ui-files/AutoscoperMainWindow.ui:399]: Thanks for looking into this. After reviewing remaining of `AutoscoperMainWindow.ui`, proposed approach is consistent with how other `QToolButton` are setup

**Review by jcfr — APPROVED** [review]

**Pull request merged**